protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,965 @@
+    ///       issues `ClosestDisjointPeersIter` tries to return the
+    ///       `num_results` closest benign peers, but as it can not
+    ///       differentiate benign from faulty paths it as well returns faulty
+    ///       peers and thus overall returns more than `num_results` peers.
+    pub fn into_result(self) -> impl Iterator<Item = PeerId> {
+        let result_per_path= self.iters.into_iter()
+            .map(|iter| iter.into_result().map(Key::from));
+
+        ResultIter::new(self.target, result_per_path).map(Key::into_preimage)
+    }
+}
+
+/// Index into the [`ClosestDisjointPeersIter`] `iters` vector.
+#[derive(Copy, Clone, Debug)]
+struct IteratorIndex(usize);
+
+impl From<usize> for IteratorIndex {
+    fn from(i: usize) -> Self {
+        IteratorIndex(i)
+    }
+}
+
+impl From<IteratorIndex> for usize {
+    fn from(i: IteratorIndex) -> Self {
+        i.0
+    }
+}
+
+impl Add<usize> for IteratorIndex {
+    type Output = Self;
+
+    fn add(self, rhs: usize) -> Self::Output {
+        (self.0 + rhs).into()
+    }
+}
+
+impl Index<IteratorIndex> for Vec<ClosestPeersIter> {
+    type Output = ClosestPeersIter;
+
+    fn index(&self, index: IteratorIndex) -> &Self::Output {
+        &self[index.0]
+    }
+}
+
+impl IndexMut<IteratorIndex> for Vec<ClosestPeersIter> {
+    fn index_mut(&mut self, index: IteratorIndex) -> &mut Self::Output {
+        &mut self[index.0]
+    }
+}
+
+/// State tracking the iterator that yielded (i.e. tried to contact) a peer. See
+/// [`ClosestDisjointPeersIter::on_success`] for details.
+#[derive(Debug)]
+struct PeerState {
+    /// First iterator to yield the peer. Will be notified both of the outcome
+    /// (success/failure) as well as the closer peers.
+    initiated_by: IteratorIndex,
+    /// Keeping track of the response state. In case other iterators later on
+    /// yield the same peer, they can be notified of the response outcome.
+    response: ResponseState,
+}
+
+impl PeerState {
+    fn new(initiated_by: IteratorIndex) -> Self {
+        PeerState {
+            initiated_by,
+            response: ResponseState::Waiting,
+        }
+    }
+}
+
+#[derive(Debug)]
+enum ResponseState {
+    Waiting,
+    Succeeded,
+    Failed,
+}
+
+/// Iterator combining the result of multiple [`ClosestPeersIter`] into a single
+/// deduplicated ordered iterator.
+//
+// Note: This operates under the assumption that `I` is ordered.
+#[derive(Clone, Debug)]
+struct ResultIter<I> where
+    I: Iterator<Item = Key<PeerId>>,
+{
+    target: KeyBytes,
+    iters: Vec<Peekable<I>>,
+}
+
+impl<I: Iterator<Item = Key<PeerId>>> ResultIter<I> {
+    fn new(target: KeyBytes, iters: impl Iterator<Item = I>) -> Self {
+        ResultIter{
+            target,
+            iters: iters.map(Iterator::peekable).collect(),
+        }
+    }
+}
+
+impl<I: Iterator<Item = Key<PeerId>>> Iterator for ResultIter<I> {
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
This is wrong, it doesn't ensure the results are from separate peers. If you have result iterables like this:

1. [1001, 1002, 1003, 1004, 1005]
2. [1010, 1020, 1030]
3. [1011, 1021]

the first 5 results will all be from the first peer, admitting an eclipse attack.

Reply:
Furthemore, I don't think it's possible to get a correct solution with the "disjoint paths" approach, unless you implement something equivalent to max-flow here, and then you might as well ditch the whole "disjoint paths" approach and do max-flow for the whole algorithm (i.e. not attempt to compose it out of separate `ClosestPeerIter`s)

Reply:
> This is wrong, it doesn't ensure the results are from separate peers. If you have result iterables like this:
> 
>     1. [1001, 1002, 1003, 1004, 1005]
> 
>     2. [1010, 1020, 1030]
> 
>     3. [1011, 1021]
> 
> 
> the first 5 results will all be from the first peer, admitting an eclipse attack.

I don't think this is correct.

In the case of a *GET_VALUE* query we make sure that there is at least one record from each path in the final result set.

In the case of a *FIND_NODE* query we make sure that we include 20 nodes of each disjoint path in the final result set.

In the case of a *PUT_VALUE* query we make sure that the record is stored on at least one node of each disjoint path.

Reply:
> This is wrong, it doesn't ensure the results are from separate peers.

It is not the intention of this implementation to ensure the first `n` results are from separate peers. This result iterator just combines the results from all paths, it does not strip off any results, we return them all (and they are all used for a subsequent `STORE` operation, if this query was the preparation phase for a `put_record`). Since you don't know which result peers are benign and which aren't, regardless of the algorithm used, this (i.e. not to discard any results) is what we want to do.

Reply:
> [..] we make sure that there is at least one record from each path in the final result set [..]

How does that possibly happen? The consumer of the iterator has no idea of what "paths" there are - this result iterator (this function) hides this information in the merging, and does not preserve this property, as per my example directly above.

> It is not the intention of this implementation to ensure the first n results are from separate peers.

It should be, and it is wrong not to have that intention: it is the intention of the general approach of "disjoint paths" to ensure that the results are from separate peers, because this protects against an eclipse attack.

This `ResultIter` implementation voids all the benefits of the disjoint paths enforcement in the other parts of the code, and an eclipse attack again is trivial - the first peer can return fake "close" results, and cause the first iterator to succeed with these, then this `ResultIter` will return these and ignore everything else.

So it is really pointless to do it this way, you are not gaining any additional security on top of `ClosestPeerIter`, an eclipse attack is as easy as before.

Furthermore, a max-flow-based solution is the theoretically-sound version of the "disjoint paths" approach, which gracefully handles (without any special logic) the examples mentioned previously in this PR, that led up to this `ResultIter` special-casing.

Reply:
> > [..] we make sure that there is at least one record from each path in the final result set [..]

> How does that possibly happen? The consumer of the iterator has no idea of what "paths" there are - this result iterator (this function) hides this information in the merging, and does not preserve this property, as per my example directly above.

This is happening here:

https://github.com/libp2p/rust-libp2p/blob/ea7f3ede68aab24a27032d44050c1c8ec806b89f/protocols/kad/src/behaviour.rs#L1519-L1527

When receiving a record from a remote peer we first of all check whether we reached the configured `Qurorum`. If we reached the quorum we tell the `query` that it may finish, unless it has objections.

The implementation of `may_finish` of a `ClosestDisjointPathsIterator` makes sure that it only finished if each path has a peer that returned a record. This is happening here:

https://github.com/libp2p/rust-libp2p/blob/ea7f3ede68aab24a27032d44050c1c8ec806b89f/protocols/kad/src/query/peers/closest/disjoint.rs#L272-L287

Does that make sense @infinity0?

Reply:
> > It is not the intention of this implementation to ensure the first n results are from separate peers.

> It should be, and it is wrong not to have that intention: it is the intention of the general approach of "disjoint paths" to ensure that the results are from separate peers, because this protects against an eclipse attack.

@infinity0 a user of the `ResultIter` is expected to fully consume the iterator, thus the user will end up with close to 20 peers from each disjoint path.